source/adaptive_load/metrics_plugin_impl.cc

```diff
@@ -0,0 +1,120 @@
+#include "adaptive_load/metrics_plugin_impl.h"
+
+#include "envoy/registry/registry.h"
+
+#include "external/envoy/source/common/common/assert.h"
+#include "external/envoy/source/common/protobuf/protobuf.h"
+
+#include "absl/status/status.h"
+#include "absl/strings/numbers.h"
+
+namespace Nighthawk {
+
+NighthawkStatsEmulatedMetricsPlugin::NighthawkStatsEmulatedMetricsPlugin(
+    const nighthawk::client::Output& nighthawk_output) {
+  errors_ = "'global' result not found in Nighthawk output.\n";
+
+  for (const nighthawk::client::Result& result : nighthawk_output.results()) {
+    if (result.name() != "global") {
+      continue;
+    }
+    errors_ = "";
+    const int64_t actual_duration_seconds =
+        Envoy::Protobuf::util::TimeUtil::DurationToSeconds(result.execution_duration());
+    const uint32_t number_of_workers =
+        nighthawk_output.results_size() == 1 ? 1 : nighthawk_output.results_size() - 1;
+    const long total_specified = nighthawk_output.options().requests_per_second().value() *
+                                 actual_duration_seconds * number_of_workers;
+    int total_sent = -1;
+    int total_2xx = -1;
+    for (const nighthawk::client::Counter& counter : result.counters()) {
+      if (counter.name() == "benchmark.http_2xx") {
+        total_2xx = counter.value();
+      } else if (counter.name() == "upstream_rq_total") {
+        total_sent = counter.value();
+      }
+    }
+    if (total_2xx < 0) {
+      errors_ += "Counter 'benchmark.total_2xx' not found.\n";
+    }
+    if (total_sent < 0) {
+      errors_ += "Counter 'upstream_rq_total' not found.\n";
+    }
+    if (actual_duration_seconds > 0.0) {
+      metric_from_name_["attempted-rps"] =
+          static_cast<double>(total_specified) / actual_duration_seconds;
```

Comment:
Maybe I'm missing something really obvious. Why do we have a static cast to double for something we just declared as long higher up? Why not just declare it as a double if we want a double?

Reply:
I was thinking of the calculation being clearer if we compute the result as the type it really is, then cast it to double only when forced by the map that expects doubles.

Reply:
I guess so, but it feels a little weird when we only ever use it as a double.

Reply:
I moved all the double casts to the time of initialization, and it does seem simpler now.